src/control/server/instance_drpc.go

```diff
@@ -75,3 +79,92 @@ func (srv *IOServerInstance) CallDrpc(module, method int32, body proto.Message)
 
 	return makeDrpcCall(dc, module, method, body)
 }
+
+// drespToMemberResult converts drpc.Response to system.MemberResult.
+//
+// MemberResult is populated with rank, state and error dependent on processing
+// dRPC response. Target state param is populated on success, Errored otherwise.
+func drespToMemberResult(rank system.Rank, action string, dresp *drpc.Response, err error, tState system.MemberState) *system.MemberResult {
+	var outErr error
+	state := system.MemberStateErrored
+
+	if err != nil {
+		outErr = errors.WithMessagef(err, "rank %s dRPC failed", &rank)
+	} else {
+		resp := &mgmtpb.DaosResp{}
+		if err = proto.Unmarshal(dresp.Body, resp); err != nil {
+			outErr = errors.WithMessagef(err, "rank %s dRPC unmarshal failed",
+				&rank)
+		} else if resp.GetStatus() != 0 {
+			outErr = errors.Errorf("rank %s dRPC returned DER %d",
+				&rank, resp.GetStatus())
+		}
+	}
+
+	if outErr == nil {
+		state = tState
+	}
+
+	return system.NewMemberResult(rank, action, outErr, state)
+}
+
+// dPing attempts dRPC request to given rank managed by instance and return
+// success or error from call result or timeout encapsulated in result.
+func (srv *IOServerInstance) dPing(ctx context.Context) *system.MemberResult {
+	rank, err := srv.GetRank()
+	if err != nil {
+		return nil // no rank to return result for
+	}
+
+	if !srv.isReady() {
+		return system.NewMemberResult(rank, "ping", nil, system.MemberStateStopped)
+	}
+
+	resChan := make(chan *system.MemberResult)
+	go func() {
+		dresp, err := srv.CallDrpc(drpc.ModuleMgmt, drpc.MethodPingRank, nil)
+		resChan <- drespToMemberResult(rank, "ping", dresp, err, system.MemberStateReady)
+	}()
+
+	select {
+	case <-ctx.Done():
+		if ctx.Err() == context.DeadlineExceeded {
```

Comment:
Does the context show up with a timeout set by default? Or am I just missing where it's set for this command? Seems really important for a ping

Reply:
this is set in the caller